i3: fix and reenable the testsuite

sigprof opened the pull request:
## Description of changes

Although the `i3` package has an extensive testsuite, the execution of that testsuite was disabled since #8399 (the bug report about failing tests was #7957).  The major problem which prevented the execution of the testsuite was that `perlPackages.X11XCB` was broken; that was fixed in #294728, and this PR fixes the rest of problems in the `i3` package and enables the testsuite again.

Changes done:
1. The `checkPhase` script was stale and needed to be rewritten for the new version of i3 (paths are different now, and `complete-run.pl` now invokes `xvfb-run` internally).  The code to check the log file for errors might be unneeded for the new version (`complete-run.pl` seems to return a non-zero exit code correctly on errors), but is left to catch any possible regressions in the test runner behavior.
2. The `318-i3-dmenu-desktop.t` testcase was failing, because that testcase was creating a temporary Perl script intended to shadow the real `i3-msg` executable, but the `#!/usr/bin/env perl` shebang in that script did not work in the build environment; this problem was not really obvious, because `system('i3-msg', $cmd)` silently continued to search for the `i3-msg` executable further in `$PATH` and launched the real binary instead of the replacement script.  The problematic shebang needed to be replaced manually (using `substituteInPlace` in `postPatch`), because `patchShebangs` handles only real shebangs on the first line of each executable file.
3. Some tests required more dependencies, which needed to be added to `buildInputs`:
   - `xdotool`
   - `xorg.setxkbmap`
   - `xorg.xrandr`
   - `which`
  
   Unfortunately, not having these packages in `buildInputs` does not actually cause test failures, because the corresponding testcases check for the availability of those tools and skip the actual tests if the tools are not found; so if any future updates of `i3` would add more tests with such optional dependencies, this may be missed.  But detecting this problem automatically is hard, because the testsuite also has other tests which get skipped (e.g., there are some tests which were written before the corresponding behavior had actually been implemented in `i3`), so just the presence of `skip` marks in the test log does not mean that there is a problem.
4. The `i3-rounded` package uses `i3.overrideAttrs` to build some old fork of `i3`, and apparently the testsuite in that fork is not properly maintained, and some tests are failing.  I don't intend to use that fork, but apparently the packaged commit points to the 4.20.1 release of that fork (dated 2021-11-03), while there is also the 4.21.1 release there (dated 2022-11-06), so the `i3-rounded` package does not look really maintained; anyway, I disabled tests in that package to keep it building as before.
5. The old condition to enable the testsuite (`doCheck = stdenv.hostPlatform.system == "x86_64-linux"`) was replaced with `doCheck = stdenv.isLinux`, because the tests are not `x86_64`-specific, but can be run only on Linux at the moment (the problem is that the `xvfb-run` package did not work properly on Darwin, but in #123097 it was made Linux-only to reflect that without fixing the underlying problem).
6. The condition on `checkPhase` was removed (it was introduced in #6135 to fix the evaluation failure on Darwin, because at that time the `checkPhase` script contained `${xdummy}`, and the `xdummy` package was not available on Darwin; however, the current version of the `checkPhase` script does not have any references to package paths and therefore does not need a conditional guard).
7. The packages which are required only for running the testsuite are added to `buildInputs` only if the platform actually supports running the testsuite (reusing the `doCheck` condition).  This change makes the `i3` package buildable on Darwin (apparently someone actually wanted that and fixed the build in #6135, but nobody stepped up to fix the Darwin build after #123097 broke it).  However, I can't actually test the Darwin build apart from looking at the ofborg logs.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

sigprof committed:
i3: fix and reenable testsuite

The `checkPhase` script was stale and needed to be rewritten for the new
version of i3 (paths are different now, and `complete-run.pl` now
invokes `xvfb-run` internally).  The code to check the log file for
errors might be unneeded for the new version (`complete-run.pl` seems to
return a non-zero exit code correctly on errors), but is left to catch
any possible regressions in the test runner behavior.

Also the `318-i3-dmenu-desktop.t` testcase was failing, because that
testcase was creating a temporary Perl script intended to shadow the
real `i3-msg` executable, but the `#!/usr/bin/env perl` shebang in that
script did not work in the build environment; this problem was not
really obvious, because `system('i3-msg', $cmd)` silently continued to
search for the `i3-msg` executable further in `$PATH` and launched the
real binary instead of the replacement script.  The problematic shebang
needed to be replaced manually, because `patchShebangs` handles only
real shebangs on the first line of each executable file.

sigprof committed:
i3: add more dependencies used by tests

Some tests in the i3 testsuite are silently skipped if the dependencies
required to run those tests are missing.  The problematic dependencies
turned out to be `xdotool`, `xorg.setxkbmap`, `xorg.xrandr`, `which`
(the latter is used to verify the presence of some tools).  Add those
dependencies to `buildInputs`, so that all tests would be run.

Ideally such missing dependencies should be detected as test failures,
but it's hard to implement, because missing tools result just in `skip`
marks in the test log, and there are some tests which are skipped for
other reasons (e.g., tests which were written before the expected
behavior is actually implemented in i3).

sigprof committed:
i3: make postPatch work for older forks too

The `i3-rounded` package is derived from `i3`, but uses some old fork of
`i3` which does not have the `318-i3-dmenu-desktop.t` testcase.  Make
the code in `postPatch` check whether that testcase is actually present,
so that the build of `i3-rounded` does not fail.

sigprof committed:
i3-rounded: disable testsuite

Some tests in the `i3-rounded` fork are failing; disable the testsuite
to allow the package to build as before.

sigprof commented:
@stigtsp As advised in #294728, I created a PR to enable tests in the `i3` package; however, I also needed to disable tests in the [`i3-rounded`](https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/window-managers/i3/rounded.nix) package, which uses `i3.overrideAttrs` to build some old fork of `i3` — apparently the testsuite in that fork is not properly maintained, and some tests are failing.

sigprof commented:
Result of `nixpkgs-review pr 294809` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>40 packages built:</summary>
  <ul>
    <li>autotiling</li>
    <li>autotiling.dist</li>
    <li>flashfocus</li>
    <li>flashfocus.dist</li>
    <li>i3</li>
    <li>i3-balance-workspace</li>
    <li>i3-balance-workspace.dist</li>
    <li>i3-layout-manager</li>
    <li>i3-resurrect</li>
    <li>i3-resurrect.dist</li>
    <li>i3-rounded</li>
    <li>i3-rounded.debug</li>
    <li>i3-swallow</li>
    <li>i3-swallow.dist</li>
    <li>i3-wk-switch</li>
    <li>i3.debug</li>
    <li>i3a</li>
    <li>i3a.dist</li>
    <li>i3altlayout</li>
    <li>i3altlayout.dist</li>
    <li>kitti3</li>
    <li>kitti3.dist</li>
    <li>nwg-displays</li>
    <li>nwg-displays.dist</li>
    <li>nwg-panel</li>
    <li>nwg-panel.dist</li>
    <li>nwg-wrapper</li>
    <li>nwg-wrapper.dist</li>
    <li>polybarFull</li>
    <li>powerline (python311Packages.powerline)</li>
    <li>powerline.dist (python311Packages.powerline.dist)</li>
    <li>python311Packages.i3ipc</li>
    <li>python311Packages.i3ipc.dist</li>
    <li>python311Packages.py3status</li>
    <li>python311Packages.py3status.dist</li>
    <li>raiseorlaunch</li>
    <li>raiseorlaunch.dist</li>
    <li>sway-assign-cgroups</li>
    <li>sway-contrib.inactive-windows-transparency</li>
    <li>xkb-switch-i3</li>
  </ul>
</details>

stigtsp commented on pkgs/applications/window-managers/i3/default.nix:
```diff
@@ -29,27 +29,41 @@ stdenv.mkDerivation rec {
+
+    # This testcase generates a Perl executable file with a shebang, and
+    # patchShebangs can't replace a shebang in the middle of a file.
+    if [ -f testcases/t/318-i3-dmenu-desktop.t ]; then
```
Do we need to check if the test case exists?

sigprof replied:
Then I would need to add
```nix
  postPatch = ''
    patchShebangs .
  '';
```
to `i3/rounded.nix` (overwriting `oldAttrs.postPatch` instead of appending as usual) to make `i3-rounded` build again.  Would that be better?

stigtsp commented on pkgs/applications/window-managers/i3/default.nix:
```diff
@@ -29,27 +29,41 @@ stdenv.mkDerivation rec {
-  # perl-related errors later on. For more, see
-  # https://github.com/NixOS/nixpkgs/issues/7957
-  doCheck = false; # stdenv.hostPlatform.system == "x86_64-linux";
+  doCheck = stdenv.hostPlatform.system == "x86_64-linux";
```
```suggestion
  doCheck = stdenv.hostPlatform.system == "x86_64-linux";
```
Can we try enabling this for supported platforms? ofborg should allow testing on aarch64, for instance.

sigprof replied:
Looks like the `x86_64-darwin` and `aarch64-darwin` builds succeeded with those changes, and the `aarch64-linux` build succeeded after running the testsuite.

stigtsp reviewed (commented):
Builds/tests fine for me and looks pretty good in general, have some minor nits:

sigprof committed:
i3: enable the testsuite on all Linux platforms

The testsuite can run only on Linux, because it requires `xvfb-run`,
which is not available on non-Linux platforms; however, it should not be
limited just to `x86_64-linux`.

The conditional setting of `checkPhase` was introduced in #6135, because
at that time the `checkPhase` script referred to `${xdummy}`, and that
kind of reference would break evaluation if the specified package was
not available; the current version of the `checkPhase` script does not
refer to any package paths directly, therefore the condition is no
longer needed and can be removed.

sigprof committed:
i3: add test-only deps to buildInputs only when testing is enabled

This may fix the build on Darwin (the `xvfb-run` package, which is
required only to run the testsuite, is Linux-only since #123097, and the
“proper solution” mentioned in that PR was not implemented).  Apparently
someone cared about building `i3` on Darwin at the time of #6135.

fpletz commented:
Rebased onto master and fixed the merge conflict.
